### ISSUE TITLE
 function-url-no-scheme-relative 

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1084]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Good:
```css
.example {
  background: url(//www.google.com/images/example);
}
```

Bad:
```css
.example {
  background: url(http://www.google.com/images/example);
}
```

```css
.example {
  background: url(https://www.google.com/images/example);
}
```

**Comment by Nirazul** [comment]
I'm not sure which way it should be enforced as your "good" example is described as an anti-pattern:
[Reference](http://www.paulirish.com/2010/the-protocol-relative-url/)

**Comment by ntwb** [comment]
> I'm not sure which way it should be enforced as your "good" example is described as an anti-pattern:

I agree, you should no longer use `//`, use `http://` or `https://`

Wondering if enforcing `http://` or `https://` could be included in the [function-url-data-uris](https://github.com/stylelint/stylelint/tree/master/src/rules/function-url-data-uris) rule with a name change to `function-url-uris`?

**Comment by jeddy3** [comment]
> Wondering if enforcing http:// or https:// could be included in the function-url-data-uris rule with a name change to function-url-uris?

Interesting suggestion. I'm wondering if offering two rules provides more flexibility though. Given `function-url-protocol: "http"|"https"|"relative"` (which keeps `relative` available as an option as these things tend to go around and around), you could have:

```json
{ 
  "rules": {
     "function-url-data-uris": "never",
     "function-url-protocol": "https"
   }
}
```

i.e. Disallow data URIs in source files and, *if the url contains a protocol*, it must be `https`. This configuration *allows for relative and absolute paths* to assets.

Another example is how `function-url-data-uris` can be used to enforce *only* data URIs are used (i.e. no protocols and no relative/absolute paths):

```json
{ 
  "rules": {
     "function-url-data-uris": "always"
   }
}
```

I think this is something that would be difficult (or at least confusing) to do if we merged this proposal into `function-url-data-uris`.

**Comment by evilebottnawi** [comment]
@jeddy3 sound good. Relative with domain or without domain (with protocol or without)?

**Comment by ntwb** [comment]
> So my suggestion, is a new rule: function-url-protocol: "http"|"https"|"relative". Thoughts?

Nice @jeddy3, this looks great 👍

**Comment by davidtheclark** [comment]
I was thinking from the discussion above that maybe we need `function-url-no-relative-protocol`. The asset itself might determine whether you use `http` or `https`, so why would you want to lock in to one or the other?

**Comment by evilebottnawi** [comment]
@davidtheclark sometimes i need to use only relative. Be good have generic rule for this. May be `function-url-format`: "relative" | "absolute", and `options`: { "protocol": true | false }. Plus this will then add another option if you need it.

**Comment by davidtheclark** [comment]
I think those should be separate rules. Because I may not want to *always* use absolute protocols, but when I do I want to *always* include the protocol.

**Comment by jeddy3** [comment]
> The asset itself might determine whether you use http or https, so why would you want to lock in to one or the other?

I was thinking of this use case: a team prefers to use secure assets i.e. those over https. If a secure version of the asset isn't available then a team member must disable the rule with stylelint commands and provide a reason. 

> The asset itself might determine whether you use http or https

This is a good point. Perhaps adding a forth option to the proposed `function-url-protocol` rule e.g. ` "http"|"https"|"absolute"|"relative"` - where `"absolute"` is either http or https?

So, you'd use:

```json
{
  "rules": {
    "function-url-protocol": "absolute",
  }
}
```

I think, because the rule name contains "protocol", it's clear that such a configuration would allow these styles:

```css
a {
  background: url("x.jpg"); /* ignored because no protocol used */
  background: url("../x.jpg"); /* ignored because no protocol used */
  background: url("/x.jpg"); /* ignored because no protocol used */
  background: url("http://www.google.com/x.jpg");
  background: url("https://www.google.com/x.jpg");
}
```

But disallow:

```css
a {
  background: url("//www.google.com/x.jpg");
}
```

If we think the preference for secure assets isn't valid, we can just restrict the options to `"absolute"|"relative"`.

**Comment by davidtheclark** [comment]
I think it's confusing to refer to `//` as "relative" because really it's not really a relative path. Is there no better word for this?

**Comment by davidtheclark** [comment]
Eh, I guess it's a "relative" *protocol*, so fine sticking with that word if there's no better.

Wait, though: My understanding is that we don't want a `relative` option, because it's been deemed a bad practice. So we shouldn't build a rule that enforces it.

> @davidtheclark sometimes i need to use only relative.

@evilebottnawi Why do you need only relative, when it's not recommended?

I'm still kind of in favor of the simple `function-url-no-relative-protocol`.

**Comment by evilebottnawi** [comment]
@davidtheclark I think it would be good to make a rule allowing to control URL format, and the possibility set absolute and relative, and the domain and without domain perhaps is divided into a few rules

**Comment by davidtheclark** [comment]
By "relative" do you mean *actually relative paths*? Or the protocol `//` that is relative to your current protocol?

**Comment by jeddy3** [comment]
> I think it's confusing to refer to // as "relative" because really it's not really a relative path. Is there no better word for this?

Here's a [URL spec](https://url.spec.whatwg.org/#url-syntax) with lots of handy names that makes everything much clearer :)

I think we can address the "Because I may not want to always use absolute protocols, but when I do I want to always include the protocol." use case with blacklist and whitelist rules:

- `function-url-special-scheme-blacklist: []`
- `function-url-special-scheme-whitelist: []`

That just leaves use to decide how to deal with [*scheme-relative URLS*](https://url.spec.whatwg.org/#syntax-url-scheme-relative)? Either as David proposes with something like `function-url-no-scheme-relative` or a couple of different rules to cover [this syntax](https://url.spec.whatwg.org/#url-syntax).

**Comment by evilebottnawi** [comment]
@jeddy3 sounds very good

**Comment by jeddy3** [comment]
Ok, so what about:

- `function-url-special-scheme-blacklist: []` ([ref](https://url.spec.whatwg.org/#special-scheme))
- `function-url-special-scheme-whitelist: []` ([ref](https://url.spec.whatwg.org/#special-scheme))
- `function-url-path: "absolute"|"relative"` ([ref](https://url.spec.whatwg.org/#syntax-url-path-absolute) & [ref](https://url.spec.whatwg.org/#syntax-url-path-relative)) 
- `function-url-no-scheme-relative: true` ([ref](https://url.spec.whatwg.org/#syntax-url-scheme-relative))

Say you want to allow relative paths and only use https when a scheme is used:

```json
{
  "rules": {
     "function-url-special-scheme-whitelist": ["https"],
     "function-url-no-scheme-relative": true,
     "function-url-path": "relative"
   }
}
```

Can anyone think of a use case (other than wanting to enforce *scheme-relative* urls!) not covered by this combination of rules?

**Comment by davidtheclark** [comment]
The only one I'm uncertain about is `function-url-path: "absolute"|"relative"`. I'm having some difficulty envisioning the circumstance where you can/should only allow one or the other.

**Comment by jeddy3** [comment]
> The only one I'm uncertain about is function-url-path: "absolute"|"relative". I'm having some difficulty envisioning the circumstance where you can/should only allow one or the other.

Happy to see this one pulled from the list until someone has a real-world need for it.

**Comment by jeddy3** [comment]
Lets focus on `function-url-no-scheme-relative` in this issue. We can revisit the others if some raises a need for them.

- Name: `function-url-no-scheme-relative`
- Primary option: `true`
- Secondary options: none
- Message: "Unexpected scheme-relative URL"

**Comment by jeddy3** [comment]
Lets keep things to one new rule per issue, otherwise it's going to be confusing.

I've pulled `function-url-path` into a https://github.com/stylelint/stylelint/issues/1602

**Comment by jeddy3** [comment]
Done in #1576
